### PR TITLE
FR: focus specific terminal; create new terminal with options; optionally reuse specific terminal for `runInTerminal`; 

### PR DESCRIPTION
**Pull request by frypf**
Currently, VSCode doesn't seem to offer any way of directly focusing a user-initiated terminal (ie. one that hasn't come from a task) without using `workbench.action.quickOpenTerm` (and selecting appropriately) or multiple keyboard shortcuts / mouseclicks. I've attempted to address this, and in the course of doing so hopefully opened up some additional potentially useful options, namely:
#### new command: `revealTerminal`
- focuses a specific terminal
- optionally matches by name, otherwise focuses a user-initiated terminal
- can request newest or oldest match
- creates a new terminal with given options if no match
#### new command: `createTerminal`
- creates a new terminal with given options
#### additional feature for command: `runInTerminal`
- can now optionally reuse specific terminal

BACKGROUND / RATIONALE: After coming up against microsoft/vscode#125573 I began using the `runInTerminal` action generally as a replacement for `workbench.action.terminal.sendSequence`. However there is then no direct way to run further related shell commands later on within that same terminal, without navigating specifically to it and either entering the shell commands manually or running `workbench.action.terminal.sendSequence`. To quote the OP in the closed issue above:
> This is very annoying as the whole point of these keyboard shortcuts is to avoid manually manipulating terminals.

These features attempt to provide a means to do all this within a single command / sequence / keybinding. Originally I had only sought to refactor `runInTerminal`, but along the way it made sense to allow focusing a specific terminal as a separate command. Then in due course it also seemed useful to allow creating a terminal with specified options - (this renames the new terminal after creation because the only way of differentiating a user-initiated vs task-initiated terminal AFAIK is that task-initiated terminals always have a `creationOptions.name`).

Hopefully this is of interest / use.

### CODE REVIEW

**Comment by frypf** [comment]
It offers the ability to create a terminal with a given name / icon color / cwd without specifically sending text to it immediately. I only included it originally to avoid duplicating code between `runInTerminal` and `focusTerminal`, as the ability to focus a non-task terminal without specifying a name relies on `term.creationOptions.name === undefined`. There may be other ways to initiate a named terminal directly from an in-built vscode command which would undermine this strategy but I couldn't find any. Obviously for $cwd, it's already possible to send `"cd /whatever"` to a terminal after opening it. However the icon colour I thought could be of use in particular as it allows quick visual differentiation between terminals intended for different purposes - currently the only way of creating a terminal with a different coloured icon is is to write a whole extra terminal profile and use that.

Regarding documentation, I wasn't sure what to put for a comment after `"name"` within `runInTerminal` to correctly show the difference - as it's obvious with `createTerminal` but slightly more nuanced with `runInTerminal`. As before, if you omit `"reuse"` it creates a new terminal with the given `"name"`, or you can omit`"name"` and it uses the default. But you can also omit `"name"` and include `"reuse"`, and then it matches the newest/oldest user-initiated terminal with any name. I couldn't figure a single short sentence to describe all 3 scenarios - although having read your comments on #44 I feel that's more of semantic issue and a docs section would be able to go into more detail, while the function names themselves show a different use implicitly.

Your call - if you genuinely don't see any value it could just be subsumed into `revealTerminal` and be non-user-facing.

File: src/commands/runInTerminalCommand.ts [hunk 1]
```diff
@@ -7,13 +9,17 @@ interface RunInTerminalArgs {
...
 	waitForExit?: boolean;
-	reuse?: boolean;
+	reuse?: string;
 }
 
 let onDidCloseTerminalDisposable: Disposable;
 
 export async function runInTerminalCommand(arg: RunInTerminalArgs | string): Promise<unknown> {
 	return new Promise((resolve, reject) => {
+		if (!arg) {
```
Comment (frypf) [inline src/commands/runInTerminalCommand.ts:19]: I had forgotten that I'd added this check - previously it was actually possible to create a terminal without sending text, although I don't know whether that was intentional as it wasn't documented.

**Comment by usernamehw** [comment]
Still not convinced of the need of `createTerminal` command. Can you merge all/most of the features of it into `runInTerminal`?

If `arg.text` is undefined - throw error, but if it's an empty string - create a terminal without sending text?

If any feature cannot be merged into `runInTerminal` - write another comment here explaining which one.

**Comment by frypf** [comment]
Sure I'll have a go, just think the functionality is worth keeping.

It might sit better within `revealTerminal` instead though if I refactor the expected arg schema? I can't see a need to create an empty terminal without revealing it (which would be possible via `runInTerminal`), and technically we're not _running_ anything. But the cognitive link between _revealing_ with eg. `"someSemanticallyObviousPropName": "most recent" (default) || "oldest" || "new"` seems stronger to me at least - any ideas on _someSemanticallyObviousPropName_ are most welcome though! Maybe just "focus"? Or switch it round and have the command / function name as `focusTerminal` and the prop name as "reveal"?

Anyway on a related note, I realised that my current strategy for differentiating non-task terminals falls down if `runInTerminal` is run and "reveal" isn't true - I was renaming the terminal straight after creation to be able to test against `creationOptions.name`, but `workbench.action.terminal.renameWithArg` can only act on the active terminal. So if we attempt to create a named terminal without revealing it, the name never gets updated. I'll have to have a think on an alternative strategy later 🤔.

**Comment by usernamehw** [comment]
> any ideas on someSemanticallyObviousPropName are most welcome though! Maybe just "focus"?

Maybe "target" or "which".

**Comment by frypf** [comment]
Had another crack at this today, works more robustly now.

**Comment by usernamehw** [comment]
I think codicon code should be a separate PR. And I have plans to merge [Codicon Names](https://marketplace.visualstudio.com/items?itemName=usernamehw.codicon-names) into this extension.

**Comment by usernamehw** [comment]
Although, it's probably fine. The icons don't change that often.

**Comment by frypf** [comment]
Could scrub it in favour of an generalised string - although it would make it more difficult to check that the icon in `TargetTermOpts` won't be blank. Just figured it'd round out all the available terminal creation options. I lifted it from their own reference (and included the link) - it says last updated in November.

**Comment by usernamehw** [comment]
No need to change. I'll review it this week.

**Comment by usernamehw** [comment]
"iconColor" doesn't work...

```js
"commands.focusTerminal_04": {
	"command": "commands.focusTerminal",
	"args": {
		"name": "random",
		"which": "create new",
		"icon": "calendar",
		"iconColor": "terminal.ansiRed"
	}
}
```
 It blinks with red and then - goes back to default color.

**Comment by frypf** [comment]
I can't reproduce - seems to work fine here. I tried with a variety of different `"terminal.integrated.tabs.defaultColor"` and  individual profile `"color"` settings 🤔.
Screenshot 2023-03-03 at 18 05 41

**Comment by usernamehw** [comment]
`workbench.action.terminal.changeColor` works fine.

**Comment by frypf** [comment]
I've only got access to zsh / bash, but it works fine with either.

**Comment by usernamehw** [comment]
Color is like 95% this issue https://github.com/microsoft/vscode/issues/167643

**Comment by frypf** [comment]
Weird, there's reports of it on both Mac and Windows but it must be pretty esoteric 🤔.

**Comment by usernamehw** [comment]
Looks great 👍. 

I changed "which" to "target".

**Comment by frypf** [comment]
Good stuff 👍.

Hopefully vscode fixes the issue you linked soon so people can get the full benefit - but in the meantime at least no more accidentally typing random text into files every 5 minutes when I try to focus the terminal and run shell commands after a task has run 🤞.

**Pull request merged**